README.md
# Verna

<a href="https://pypi.org/project/verna"><img alt="PyPI" src="https://img.shields.io/pypi/v/verna"></a>
<img alt="Build Status" src="https://api.travis-ci.com/ju-sh/verna.svg?branch=master"></img>
<a href="https://github.com/ju-sh/verna/blob/master/LICENSE.md"><img alt="License: MIT" src="https://img.shields.io/pypi/l/verna"></a>

A simple module to handle colors

Only RGBA colors are supported at the moment.

---

<h2>Installation</h2>

You need Python>=3.6 to use Verna.

It can be installed from PyPI with pip using

    pip install verna

---

<h2>Usage</h2>

Colors are represented using objects of class `Color`.

The color value is essentially stored as an integer as the `integer` attribute.

The following properties can be used to access the different color components.

    color.alpha
    color.red
    color.green
    color.blue

where `color` is an instance of `Color`.

The different color components can be edited with one of the following values

 - a percentage in string form with a '%' at the end (eg: "20%")
 - an integer from 0 to 255 (eg: 0xff, 255)
 - a float from 0.0 to 1.0 (eg: 0.4)

So, the following are valid:

    color = Color(0x00ffff)
    color.alpha = 0x80
    color.alpha = "50%"
    color.alpha = 0.5   # float values can be assigned
                        # only to alpha property

    color.red = 0xff    # Same as color.red = 255
    color.red = "100%"
    color.green = 217
    color.green = "85%"
    color.blue = 0xf5
    color.blue = 0.7

whereas the following will cause error:

    color = Color(0x00ffff)
    color.alpha = 0x1ff    # > 0xff
    color.alpha = -1       # < 0.0
    color.alpha = "120%"   # > 100%
    color.alpha = "120"    # No '%' at end
    color.alpha = 1.2      # > 1.0
    color.alpha = True     # Invalid type: bool

A `Color` object may be created in multiple ways.

By default, alpha value is `0`.

<h6>From integer color code</h6>

For example, cyan (solid), which has color code `0x00ffff` can be created like

    Color(0x00ffff)

which is same as

    Color(65535)

<h6>From color name</h6>

`Color.from_name()` can be used to create `Color` objects from a [CSS3 color name](https://www.w3.org/wiki/CSS3/Color/Extended_color_keywords).

For example, cyan can be created with

    Color.from_name('cyan')

<h6>From RGBA values</h6>

`Color.from_rgba()` can be used to create an instance of `Color` from RGBA values.

    Color.from_rgba(255, 255, 0)         # solid yellow
    Color.from_rgba(255, 255, 0, 0.5)    # translucent yellow



CHANGELOG.md
<h4> v0.1.0 () </h4>

 - Allow assigning float values (ranging from 0.0 to 1.0) for red,
   green and blue properties.
 - Add `to_int()`, `to_float()` and `to_percentage()` functions.

<h4> v0.0.2 (16 Oct 2020) </h4>

 - Fix error in `Color.from_name()` so that it returns `Color` instead of `int`.
 - Use setuptools for packaging.

<h4> v0.0.1 (15 Oct 2020) </h4>

First version


tests/test_verna.py
import pytest

from verna import Color


class TestEdit:
    """Check if the property values can be modified"""
    @pytest.mark.parametrize('prop,color_hex,arg', [
        ('alpha', 0xeeff22aa, 0x1ff),    # > 0xff
        ('alpha', 0xeeff22aa, -1),       # < 0.0
        ('red', 0xeeff22aa, "120%"),     # Invalid type
        ('red', 0xeeff22aa, "120"),      # Invalid type
        ('green', 0xeeff22aa, 1.2),      # > 1.0
        ('green', 0xeeff22aa, True),     # Invalid type: bool
        ('blue', 0xeeff22aa, 0x3ed),     # > 0xff
    ])
    def test_invalid_edit(self, prop, color_hex, arg):
        """Modifications that should raise exception"""
        color = Color(color_hex)
        with pytest.raises(ValueError):
            setattr(color, prop, arg)

    @pytest.mark.parametrize('prop,color_hex,arg,expected', [
        ('alpha', 0xeeff22aa, 0xff, 0xffff22aa),     # int/hex
        ('alpha', 0xeeff22aa, 0.5, 0x80ff22aa),      # float
        ('red', 0xeeff22aa, 0xed, 0xeeed22aa),
        ('green', 0xeeff22aa, 0x58, 0xeeff58aa),
        ('blue', 0xeeff22aa, 0.21, 0xeeff2236),
    ])
    def test_valid_edit(self, prop, color_hex, arg, expected):
        """Edit a color which already has an alpha value"""
        color = Color(color_hex)
        setattr(color, prop, arg)
        assert color.integer == expected


class TestPropAccess:
    """See if the different properties can be accessed"""
    @pytest.mark.parametrize('prop,color_hex,expected', [
        ('red', 0xeeff22aa, 0xff),
        ('green', 0xeeff22aa, 0x22),
        ('blue', 0xeeff22aa, 0xaa),
    ])
    def test_rgb(self, prop, color_hex, expected):
        color = Color(color_hex)
        assert getattr(color, prop) == expected

    def test_alpha(self):
        color = Color(0xeeff22aa)
        assert round(color.alpha, 2) == 0.93


class TestFromRGBA:
    """Tests for Color.from_rgba()"""
    def test_with_alpha(self):
        color = Color.from_rgba(0xef, 0xa1, 0xde, 0.5)
        assert color.integer == 0x80efa1de

    def test_without_alpha(self):
        color = Color.from_rgba(0xef, 0xa1, 0xde)
        assert color.integer == 0xefa1de


class TestDunders:
    """Test the magic methods of Color class"""
    def test_invalid_init(self):
        with pytest.raises(ValueError):
            Color(-1)

    def test_int(self):
        color = Color(0xeeff22aa)
        assert int(color) == 0xeeff22aa

    def test_repr(self):
        color = Color(0xeeff22aa)
        assert repr(color) == "<Color(4009697962)>"

    def test_str(self):
        color = Color(0xeeff22aa)
        assert str(color) == "eeff22aa"


class TestNormalize:
    """Tests for Color._normalize()"""
    @pytest.mark.parametrize('val,expected', [
        (0.5, 0.5),
        (120, 0.47058823529411764),
        (128, 0.5019607843137255),
    ])
    def test_valid(self, val, expected):
        assert Color._normalize(val) == expected

    @pytest.mark.parametrize('val', [
        "50",   # invalid type
        1.55,   # > 1.0
    ])
    def test_invalid(self, val):
        with pytest.raises(ValueError):
            assert Color._normalize(val)


class TestReplace:
    """Tests for Color.replace()"""
    @pytest.mark.parametrize('props,color_hex,expected', [
        ({'red': 0.5, 'blue': 234}, 0x1abcdef, 0x180cdea),
        ({'red': None, 'green': None, 'blue': None, 'alpha': None},
         0x1abcdef, 0x1abcdef),
        ({'red': 1, 'green': 0.7, 'blue': 181, 'alpha': 0},
         0x1abcdef, 0x0001b3b5),
        ({'red': 1.0, 'green': 0.7, 'blue': 181, 'alpha': 0},
         0x1abcdef, 0xffb3b5),
    ])
    def test_valid(self, props, color_hex, expected):
        color = Color(color_hex)
        assert color.replace(**props).integer == expected

    def test_invalid(self):
        color = Color(0x1abcdef)
        with pytest.raises(ValueError):
            color.replace(red=True)  # invalid type


def test_from_name():
    assert Color.from_name('gainsboro') == Color(0xdcdcdc)


def test_rgba():
    color = Color(0x80abcdef)
    red, green, blue, alpha = color.rgba()
    alpha = round(alpha, 2)
    assert (red, green, blue, alpha) == (0xab, 0xcd, 0xef, 0.5)


tests/test_conversions.py
import pytest

from verna import Color


class TestToInt:
    """Test Color.to_int()"""
    @pytest.mark.parametrize('val,expected', [
        (1.0, 255),
        (0.754, 192),
        (0.47, 0x78),
        (1, 1),
        (245, 245),
    ])
    def test_valid(self, val, expected):
        assert Color.to_int(val) == expected

    @pytest.mark.parametrize('val', [
        "120",  # invalid type
        11.0,   # > 1.0
        -0.32,  # < 0.0
        -32,    # < 0
        True,   # invalid type
    ])
    def test_invalid(self, val):
        with pytest.raises(ValueError):
            Color.to_int(val)


@pytest.mark.parametrize('val,expected', [
    (1.0, 1),
    (0x2e, 0.1803921568627451)
])
def test_to_float(val, expected):
    """Test Color.to_float()"""
    assert Color.to_float(val) == expected


src/verna/verna.py
"""
Data structures used to handle colors in Verna
"""

from typing import Tuple

from verna import names


def _rounded_float(val: float) -> float:
    """
    Similar to the built-in round() except that rounding is done
    towards higher value if val is equally close to the nearest values.
    Like:

    >>> _rounded_float(1.5)  # 2
    >>> _rounded_float(2.5)  # 3
    >>> _rounded_float(3.5)  # 4
    >>> _rounded_float(4.4)  # 4
    >>> _rounded_float(5.6)  # 6

    Built-in round() rounds towards an even value if val is equally close
    to both the nearest values as shown below
    >>> round(1.5)  # 2
    >>> round(2.5)  # 2
    >>> round(3.5)  # 4

    Returns int if val doesn't have decimal place values otherwise value
    returned is a float.
    """
    val_str = str(val)
    if ('.' in val_str) and (val_str[-1] == '5'):
        ndigits = len(val_str) - (val_str.find('.') + 1)
        val_str = val_str[:-1] + '6'
        return round(float(val_str), ndigits)
    return val


def _rounded_int(val: float) -> int:
    """
    Accept a float value, round it upwards with _rounded_float() and return the
    resultant value as an int.
    """
    return round(_rounded_float(val * 0xff))


class Color(int):
    """float
    Represents a color.

    Only RGB scheme is supported.
    """
    def __init__(self, integer: int):
        super().__init__()
        if integer < 0 or integer > 0xffffffff:
            raise ValueError
        self.integer = integer

    def __int__(self):
        return self.integer

    def __repr__(self):
        return f"<Color({self.integer})>"

    def __str__(self):
        """Returns core value in hex form as a string with out any prefix"""
        return hex(self.integer)[2:]

    def replace(self,
                red: float = None,
                green: float = None,
                blue: float = None,
                alpha: float = None) -> 'Color':
        """
        Returns a new Color by changing properties of a pre-existing instance.
        """
        if red is None:
            red_int = self.red
        else:
            red_int = _rounded_int(self._normalize(red))

        if green is None:
            green_int = self.green
        else:
            green_int = _rounded_int(self._normalize(green))

        if blue is None:
            blue_int = self.blue
        else:
            blue_int = _rounded_int(self._normalize(blue))

        if alpha is None:
            alpha_int = self.alpha
        else:
            alpha_int = _rounded_int(self._normalize(alpha))

        return self.__class__.from_rgba(red_int, green_int,
                                        blue_int, alpha_int)

    def rgba(self) -> Tuple[int, int, int, float]:
        """Returns red, green, blue, alpha as a tuple in that order"""
        return self.red, self.green, self.blue, self.alpha

    @classmethod
    def from_name(cls, name: str) -> 'Color':
        """
        Returns a Color object based on the given color name.

        Only CSS3 extended color keyword names are supported.
        """
        name = name.lower()
        return cls(names.COLORS[name])

    @classmethod
    def from_rgba(cls,
                  red: float,
                  green: float,
                  blue: float,
                  alpha: float = 0) -> 'Color':
        """
        Returns a Color object from a set of RGBA values
        """
        red_int = _rounded_int(cls._normalize(red))
        green_int = _rounded_int(cls._normalize(green))
        blue_int = _rounded_int(cls._normalize(blue))
        alpha_int = _rounded_int(cls._normalize(alpha))
        integer = ((alpha_int << 24) | (red_int << 16)
                   | (green_int << 8) | blue_int)
        return cls(integer)

    @classmethod
    def to_float(cls, val: float) -> float:
        """
        Returns equivalent float ranging from 0.0 to 1.0 from
        an int ranging from 0 to 255 (eg: 255 -> 1.0)
        Value of val is returned unchanged if it is a float from 0.0 to 1.0

        Raises ValueError if conversion cannot be performed.
        """
        val_float = cls._normalize(val)
        return _rounded_float(val_float)

    @classmethod
    def to_int(cls, val: float) -> int:
        """
        Returns equivalent int ranging from 0 to 255 from:
         - a float ranging from 0 to 1.0 (eg: 0.5 -> 128)
        Value of val is returned unchanged if it is an integer from 0 to 255.

        Raises ValueError if conversion cannot be performed.
        """
        val_float = cls._normalize(val)
        return _rounded_int(val_float)

    @staticmethod
    def _normalize(val: float) -> float:
        """
        Validate and convert a color component value in the form of an
        integer ranging from 0 to 255 or a float ranging from 0.0 to 1.0
        to an equivalent float ranging from 0.0 to 1.0

        val may be an int or a float.

        Raises ValueError if conversion cannot be done.

        Note: Property getter-setters of the class rely on this function.
        """

        if isinstance(val, int) and not isinstance(val, bool):
            # For int, value should be between 0 and 255 (inclusive)
            if val < 0 or val > 255:
                raise ValueError("{val}: Invalid value")
            float_val = val / 0xff

        elif isinstance(val, float):
            # For float, value should be between 0.0 and 1.0 (inclusive)
            if val < 0 or val > 1:
                raise ValueError(f"{val}: Invalid value")
            float_val = val
        else:
            raise ValueError(f"{val}: Invalid value type")

        return float_val

    @property
    def alpha(self) -> float:
        """
        Alpha value ranges from 0 to 1.0
        """
        return (self.integer >> 24) / 0xff

    @alpha.setter
    def alpha(self, val: float):
        val_int = _rounded_int(self._normalize(val))
        self.integer = (val_int << 24) | (self.integer & 0x00ffffff)

    @property
    def red(self) -> int:
        """
        Red value ranges from 0 to 255
        """
        return (self.integer >> 16) & 0xff

    @red.setter
    def red(self, val: float):
        val_int = _rounded_int(self._normalize(val))
        self.integer = (val_int << 16) | (self.integer & 0xff00ffff)

    @property
    def green(self) -> int:
        """
        Grren value ranges from 0 to 255
        """
        return (self.integer >> 8) & 0xff

    @green.setter
    def green(self, val: float):
        val_int = _rounded_int(self._normalize(val))
        self.integer = (val_int << 8) | (self.integer & 0xffff00ff)

    @property
    def blue(self) -> int:
        """
        Blue value ranges from 0 to 255
        """
        return self.integer & 0xff

    @blue.setter
    def blue(self, val: float):
        val_int = _rounded_int(self._normalize(val))
        self.integer = val_int | (self.integer & 0xffffff00)
